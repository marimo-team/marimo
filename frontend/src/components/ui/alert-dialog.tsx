/* Copyright 2024 Marimo. All rights reserved. */
import * as React from "react";
import * as AlertDialogPrimitive from "@radix-ui/react-alert-dialog";

import { cn } from "@/utils/cn";
import { buttonVariants } from "@/components/ui/button";
import { useRestoreFocus } from "./use-restore-focus";

const AlertDialog = AlertDialogPrimitive.Root;

const AlertDialogTrigger = AlertDialogPrimitive.Trigger;

const AlertDialogPortal = ({
  children,
  ...props
}: AlertDialogPrimitive.AlertDialogPortalProps) => (
  <AlertDialogPrimitive.Portal {...props}>
    <div className="fixed inset-0 z-50 flex items-end justify-center sm:items-start sm:top-[15%]">
      {children}
    </div>
  </AlertDialogPrimitive.Portal>
);
AlertDialogPortal.displayName = AlertDialogPrimitive.Portal.displayName;

const AlertDialogOverlay = React.forwardRef<
  React.ElementRef<typeof AlertDialogPrimitive.Overlay>,
  React.ComponentPropsWithoutRef<typeof AlertDialogPrimitive.Overlay>
>(({ className, children, ...props }, ref) => (
  <AlertDialogPrimitive.Overlay
    className={cn(
      "fixed inset-0 z-50 bg-background/80 backdrop-blur-sm transition-opacity animate-in fade-in",
      className,
    )}
    {...props}
    ref={ref}
  />
));
AlertDialogOverlay.displayName = AlertDialogPrimitive.Overlay.displayName;

const AlertDialogContent = React.forwardRef<
  React.ElementRef<typeof AlertDialogPrimitive.Content>,
  React.ComponentPropsWithoutRef<typeof AlertDialogPrimitive.Content>
>(({ className, ...props }, ref) => {
  // store the last focused element so we can restore it when the dialog closes
  const restoreFocus = useRestoreFocus();

  return (
    <AlertDialogPortal>
      <AlertDialogOverlay />
      <AlertDialogPrimitive.Content
        ref={ref}
        className={cn(
          "fixed z-50 grid w-full max-w-2xl scale-100 gap-4 border bg-background p-6 opacity-100 shadow-sm animate-in fade-in-90 slide-in-from-bottom-10 sm:rounded-lg sm:zoom-in-90 sm:slide-in-from-bottom-0 md:w-full",
          className,
        )}
        {...restoreFocus}
        {...props}
      />
    </AlertDialogPortal>
  );
});
AlertDialogContent.displayName = AlertDialogPrimitive.Content.displayName;

const AlertDialogHeader = ({
  className,
  ...props
}: React.HTMLAttributes<HTMLDivElement>) => (
  <div
    className={cn(
      "flex flex-col space-y-2 text-center sm:text-left",
      className,
    )}
    {...props}
  />
);
AlertDialogHeader.displayName = "AlertDialogHeader";

const AlertDialogFooter = ({
  className,
  ...props
}: React.HTMLAttributes<HTMLDivElement>) => (
  <div
    className={cn(
      "flex flex-col-reverse sm:flex-row sm:justify-end sm:space-x-2",
      className,
    )}
    {...props}
  />
);
AlertDialogFooter.displayName = "AlertDialogFooter";

const AlertDialogTitle = React.forwardRef<
  React.ElementRef<typeof AlertDialogPrimitive.Title>,
  React.ComponentPropsWithoutRef<typeof AlertDialogPrimitive.Title>
>(({ className, ...props }, ref) => (
  <AlertDialogPrimitive.Title
    ref={ref}
    className={cn("text-lg font-semibold", className)}
    {...props}
  />
));
AlertDialogTitle.displayName = AlertDialogPrimitive.Title.displayName;

const AlertDialogDescription = React.forwardRef<
  React.ElementRef<typeof AlertDialogPrimitive.Description>,
  React.ComponentPropsWithoutRef<typeof AlertDialogPrimitive.Description>
>(({ className, ...props }, ref) => (
  <AlertDialogPrimitive.Description
    ref={ref}
    className={cn("text-muted-foreground", className)}
    {...props}
  />
));
AlertDialogDescription.displayName =
  AlertDialogPrimitive.Description.displayName;

const AlertDialogAction = React.forwardRef<
  React.ElementRef<typeof AlertDialogPrimitive.Action>,
  React.ComponentPropsWithoutRef<typeof AlertDialogPrimitive.Action>
>(({ className, ...props }, ref) => (
  <AlertDialogPrimitive.Action
    ref={ref}
    className={cn(buttonVariants(), className)}
    {...props}
  />
));
AlertDialogAction.displayName = AlertDialogPrimitive.Action.displayName;

const AlertDialogDestructiveAction = React.forwardRef<
  React.ElementRef<typeof AlertDialogPrimitive.Action>,
  React.ComponentPropsWithoutRef<typeof AlertDialogPrimitive.Action>
>(({ className, ...props }, ref) => (
  <AlertDialogPrimitive.Action
    ref={ref}
    className={cn(
      buttonVariants({
        variant: "destructive",
      }),
      className,
    )}
    {...props}
  />
));
AlertDialogDestructiveAction.displayName = "AlertDialogDestructiveAction";

const AlertDialogCancel = React.forwardRef<
  React.ElementRef<typeof AlertDialogPrimitive.Cancel>,
  React.ComponentPropsWithoutRef<typeof AlertDialogPrimitive.Cancel>
>(({ className, ...props }, ref) => (
  <AlertDialogPrimitive.Cancel
    ref={ref}
    className={cn(
      buttonVariants({ variant: "secondary" }),
      "mt-2 sm:mt-0",
      className,
    )}
    {...props}
  />
));
AlertDialogCancel.displayName = AlertDialogPrimitive.Cancel.displayName;

export {
  AlertDialog,
  AlertDialogTrigger,
  AlertDialogContent,
  AlertDialogHeader,
  AlertDialogFooter,
  AlertDialogTitle,
  AlertDialogDescription,
  AlertDialogAction,
  AlertDialogDestructiveAction,
  AlertDialogCancel,
};
