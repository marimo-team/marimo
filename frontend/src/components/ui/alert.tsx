/* Copyright 2024 Marimo. All rights reserved. */
import * as React from "react";
import { VariantProps, cva } from "class-variance-authority";

import { cn } from "@/utils/cn";

const alertVariants = cva(
  "relative w-full rounded-lg border p-4 [&>svg]:absolute [&>svg]:text-foreground [&>svg]:left-4 [&>svg]:top-4 [&>svg+div]:translate-y-[-3px] [&:has(svg)]:pl-11",
  {
    variants: {
      variant: {
        default: "bg-background text-foreground",
        destructive:
          "text-destructive border-destructive/50 dark:border-destructive [&>svg]:text-destructive text-destructive",
      },
    },
    defaultVariants: {
      variant: "default",
    },
  },
);

const Alert = React.forwardRef<
  HTMLDivElement,
  React.HTMLAttributes<HTMLDivElement> & VariantProps<typeof alertVariants>
>(({ className, variant, ...props }, ref) => (
  <div
    ref={ref}
    role="alert"
    className={cn(alertVariants({ variant }), className)}
    {...props}
  />
));
Alert.displayName = "Alert";

const AlertTitle = React.forwardRef<
  HTMLParagraphElement,
  React.HTMLAttributes<HTMLHeadingElement>
>(({ className, children, ...props }, ref) => (
  <h5
    ref={ref}
    className={cn("mb-1 font-medium leading-none tracking-tight", className)}
    {...props}
  >
    {children}
  </h5>
));
AlertTitle.displayName = "AlertTitle";

const AlertDescription = React.forwardRef<
  HTMLParagraphElement,
  React.HTMLAttributes<HTMLParagraphElement>
>(({ className, ...props }, ref) => (
  <div
    ref={ref}
    className={cn("text-sm [&_p]:leading-relaxed", className)}
    {...props}
  />
));
AlertDescription.displayName = "AlertDescription";

export { Alert, AlertTitle, AlertDescription };
