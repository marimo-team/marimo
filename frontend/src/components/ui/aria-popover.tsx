/* Copyright 2024 Marimo. All rights reserved. */
import { cn } from "@/utils/cn";
import {
  Dialog as AriaDialog,
  type DialogProps as AriaDialogProps,
  DialogTrigger as AriaDialogTrigger,
  Popover as AriaPopover,
  type PopoverProps as AriaPopoverProps,
  composeRenderProps,
} from "react-aria-components";

const PopoverTrigger = AriaDialogTrigger;

const Popover = ({ className, offset = 4, ...props }: AriaPopoverProps) => (
  <AriaPopover
    offset={offset}
    className={composeRenderProps(className, (className) =>
      cn(
        "z-50 rounded-md border bg-popover text-popover-foreground shadow-md outline-none",
        /* Entering */
        "data-[entering]:animate-in data-[entering]:fade-in-0 data-[entering]:zoom-in-95",
        /* Exiting */
        "data-[exiting]:animate-out data-[exiting]:fade-out-0 data-[exiting]:zoom-out-95",
        /* Placement */
        "data-[placement=bottom]:slide-in-from-top-2 data-[placement=left]:slide-in-from-right-2 data-[placement=right]:slide-in-from-left-2 data-[placement=top]:slide-in-from-bottom-2",
        className,
      ),
    )}
    {...props}
  />
);

const PopoverDialog = ({ className, ...props }: AriaDialogProps) => {
  return (
    <AriaDialog className={cn("p-4 outline outline-0", className)} {...props} />
  );
};

export { Popover, PopoverTrigger, PopoverDialog };
