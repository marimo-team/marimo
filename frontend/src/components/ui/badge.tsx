/* Copyright 2024 Marimo. All rights reserved. */
import * as React from "react";
import { VariantProps, cva } from "class-variance-authority";

import { cn } from "@/utils/cn";

const badgeVariants = cva(
  "inline-flex items-center border rounded-full px-2 py-0.5 text-xs font-semibold transition-colors focus:outline-none focus:ring-2 focus:ring-ring focus:ring-offset-2",
  {
    variants: {
      variant: {
        default:
          "bg-primary hover:bg-primary/80 border-transparent text-primary-foreground",
        secondary:
          "bg-secondary hover:bg-secondary/80 border-transparent text-secondary-foreground",
        destructive:
          "bg-[var(--red-2)] border-[var(--red-6)] text-[var(--red-9)] hover:bg-[var(--red-3)]",
        success:
          "bg-[var(--grass-2)] border-[var(--grass-5)] text-[var(--grass-9)] hover:bg-[var(--grass-3)]",
        outline: "text-foreground",
      },
    },
    defaultVariants: {
      variant: "default",
    },
  },
);

export interface BadgeProps
  extends React.HTMLAttributes<HTMLDivElement>,
    VariantProps<typeof badgeVariants> {}

const Badge = ({ className, variant, ...props }: BadgeProps) => {
  return (
    <div className={cn(badgeVariants({ variant }), className)} {...props} />
  );
};

export { Badge, badgeVariants };
