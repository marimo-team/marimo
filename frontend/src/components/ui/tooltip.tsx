/* Copyright 2024 Marimo. All rights reserved. */
import * as React from "react";
import * as TooltipPrimitive from "@radix-ui/react-tooltip";

import { cn } from "@/utils/cn";

const TooltipProvider = TooltipPrimitive.Provider;

const TooltipRoot = TooltipPrimitive.Root;

const TooltipTrigger = TooltipPrimitive.Trigger;

const TooltipContent = React.forwardRef<
  React.ElementRef<typeof TooltipPrimitive.Content>,
  React.ComponentPropsWithoutRef<typeof TooltipPrimitive.Content>
>(({ className, sideOffset = 4, ...props }, ref) => (
  <TooltipPrimitive.Content
    ref={ref}
    sideOffset={sideOffset}
    className={cn(
      "z-50 overflow-hidden rounded-md border bg-popover px-3 py-1.5 text-sm text-popover-foreground shadow-xs data-[side=bottom]:slide-in-from-top-1 data-[side=left]:slide-in-from-right-1 data-[side=right]:slide-in-from-left-1 data-[side=top]:slide-in-from-bottom-1",
      className,
    )}
    {...props}
  />
));
TooltipContent.displayName = TooltipPrimitive.Content.displayName;

const Tooltip: React.FC<
  {
    content: React.ReactNode;
    usePortal?: boolean;
    children: React.ReactNode;
    side?: TooltipPrimitive.TooltipContentProps["side"];
  } & React.ComponentPropsWithoutRef<typeof TooltipRoot>
> = ({ content, children, usePortal = true, side, ...rootProps }) => (
  <TooltipRoot disableHoverableContent={true} {...rootProps}>
    <TooltipTrigger asChild={true}>{children}</TooltipTrigger>
    {usePortal ? (
      <TooltipPrimitive.TooltipPortal>
        <TooltipContent side={side}>{content}</TooltipContent>
      </TooltipPrimitive.TooltipPortal>
    ) : (
      <TooltipContent side={side}>{content}</TooltipContent>
    )}
  </TooltipRoot>
);

export {
  Tooltip,
  TooltipRoot,
  TooltipTrigger,
  TooltipContent,
  TooltipProvider,
};
