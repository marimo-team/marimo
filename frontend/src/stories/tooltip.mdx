import { Meta, Story } from "@storybook/blocks";

import { Plus } from "lucide-react";

import { Button } from "@/components/ui/button";
import {
  TooltipRoot,
  TooltipProvider,
  TooltipContent,
  TooltipTrigger,
} from "@/components/ui/tooltip";

<Meta title="Tooltip" component={TooltipRoot} />

# Tooltip

<TooltipProvider>
  <TooltipRoot>
    <TooltipTrigger asChild>
      <Button variant="outline" className="rounded-full">
        <Plus className="h-4 w-4" />

        <span className="sr-only">Add</span>
      </Button>
    </TooltipTrigger>

    <TooltipContent>
      <p>Add to library</p>
    </TooltipContent>

  </TooltipRoot>
</TooltipProvider>
