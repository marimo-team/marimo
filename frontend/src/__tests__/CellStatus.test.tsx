/* Copyright 2024 Marimo. All rights reserved. */
import {
  CellStatusComponent,
  ElapsedTime,
  formatElapsedTime,
} from "../components/editor/cell/CellStatus";
import { afterEach, beforeEach, describe, expect, test, vi } from "vitest";
import { render } from "@testing-library/react";
import type { Seconds } from "@/utils/time";
import { TooltipProvider } from "@/components/ui/tooltip";

// Mock date-fns to have consistent date formatting in tests
vi.mock("date-fns", () => ({
  formatDistanceToNow: () => "5 minutes",
}));

describe("formatElapsedTime", () => {
  test("formats milliseconds correctly", () => {
    expect(formatElapsedTime(500)).toBe("500ms");
    expect(formatElapsedTime(50)).toBe("50ms");
  });

  test("formats seconds correctly", () => {
    expect(formatElapsedTime(1500)).toBe("1.50s");
    expect(formatElapsedTime(2340)).toBe("2.34s");
  });

  test("formats minutes and seconds correctly", () => {
    expect(formatElapsedTime(60 * 1000)).toBe("1m0s");
    expect(formatElapsedTime(90 * 1000)).toBe("1m30s");
    expect(formatElapsedTime(89 * 1000)).toBe("1m29s");
    expect(formatElapsedTime(91 * 1000)).toBe("1m31s");
    expect(formatElapsedTime(150 * 1000)).toBe("2m30s");
    expect(formatElapsedTime(151 * 1000)).toBe("2m31s");
  });

  test("handles null input", () => {
    expect(formatElapsedTime(null)).toBe("");
  });
});

describe("ElapsedTime component", () => {
  test("renders elapsed time correctly", () => {
    const { container } = render(
      <TooltipProvider>
        <ElapsedTime elapsedTime="1.50s" />
      </TooltipProvider>,
    );
    expect(container).toMatchSnapshot();
  });
});

describe("CellStatusComponent", () => {
  // Mock date for consistent testing
  const mockDate = new Date(2024, 0, 1, 12, 0, 0);
  const originalDate = global.Date;

  beforeEach(() => {
    global.Date = class extends Date {
      static override now() {
        return mockDate.getTime();
      }
    } as typeof Date;
  });

  afterEach(() => {
    global.Date = originalDate;
  });

  // Base props that will be modified for different test cases
  const baseProps = {
    editing: true,
    edited: false,
    disabled: false,
    staleInputs: false,
    status: "idle" as const,
    interrupted: false,
    elapsedTime: null,
    runStartTimestamp: null,
    lastRunStartTimestamp: null,
    uninstantiated: false,
  };

  test("returns null when not editing", () => {
    const { container } = render(
      <TooltipProvider>
        <CellStatusComponent {...baseProps} editing={false} />
      </TooltipProvider>,
    );
    expect(container.firstChild).toBeNull();
  });

  test("renders disabled and stale state", () => {
    const props = {
      ...baseProps,
      disabled: true,
      staleInputs: true,
      lastRunStartTimestamp: 1_704_096_000 as Seconds, // Jan 1, 2024 12:00:00
    };

    const { container } = render(
      <TooltipProvider>
        <CellStatusComponent {...props} />
      </TooltipProvider>,
    );
    expect(container).toMatchSnapshot();
  });

  test("renders disabled state", () => {
    const props = {
      ...baseProps,
      disabled: true,
      lastRunStartTimestamp: 1_704_096_000 as Seconds,
    };

    const { container } = render(
      <TooltipProvider>
        <CellStatusComponent {...props} />
      </TooltipProvider>,
    );
    expect(container).toMatchSnapshot();
  });

  test("renders disabled transitively state", () => {
    const props = {
      ...baseProps,
      status: "disabled-transitively" as const,
      lastRunStartTimestamp: 1_704_096_000 as Seconds,
    };

    const { container } = render(
      <TooltipProvider>
        <CellStatusComponent {...props} />
      </TooltipProvider>,
    );
    expect(container).toMatchSnapshot();
  });

  test("renders stale and disabled transitively state", () => {
    const props = {
      ...baseProps,
      status: "disabled-transitively" as const,
      staleInputs: true,
      lastRunStartTimestamp: 1_704_096_000 as Seconds,
    };

    const { container } = render(
      <TooltipProvider>
        <CellStatusComponent {...props} />
      </TooltipProvider>,
    );
    expect(container).toMatchSnapshot();
  });

  test("renders running state", () => {
    const props = {
      ...baseProps,
      status: "running" as const,
      runStartTimestamp: 1_704_096_000 as Seconds,
      lastRunStartTimestamp: 1_704_096_000 as Seconds,
    };

    const { container } = render(
      <TooltipProvider>
        <CellStatusComponent {...props} />
      </TooltipProvider>,
    );
    expect(container).toMatchSnapshot();
  });

  test("renders queued state", () => {
    const props = {
      ...baseProps,
      status: "queued" as const,
      lastRunStartTimestamp: 1_704_096_000 as Seconds,
    };

    const { container } = render(
      <TooltipProvider>
        <CellStatusComponent {...props} />
      </TooltipProvider>,
    );
    expect(container).toMatchSnapshot();
  });

  test("renders uninstantiated state", () => {
    const props = {
      ...baseProps,
      uninstantiated: true,
    };

    const { container } = render(
      <TooltipProvider>
        <CellStatusComponent {...props} />
      </TooltipProvider>,
    );
    expect(container).toMatchSnapshot();
  });

  test("renders interrupted state", () => {
    const props = {
      ...baseProps,
      interrupted: true,
      elapsedTime: 1500,
      lastRunStartTimestamp: 1_704_096_000 as Seconds,
    };

    const { container } = render(
      <TooltipProvider>
        <CellStatusComponent {...props} />
      </TooltipProvider>,
    );
    expect(container).toMatchSnapshot();
  });

  test("renders edited state", () => {
    const props = {
      ...baseProps,
      edited: true,
      elapsedTime: 1500,
      lastRunStartTimestamp: 1_704_096_000 as Seconds,
    };

    const { container } = render(
      <TooltipProvider>
        <CellStatusComponent {...props} />
      </TooltipProvider>,
    );
    expect(container).toMatchSnapshot();
  });

  test("renders stale inputs state", () => {
    const props = {
      ...baseProps,
      staleInputs: true,
      elapsedTime: 1500,
      lastRunStartTimestamp: 1_704_096_000 as Seconds,
    };

    const { container } = render(
      <TooltipProvider>
        <CellStatusComponent {...props} />
      </TooltipProvider>,
    );
    expect(container).toMatchSnapshot();
  });

  test("renders completed run with elapsed time", () => {
    const props = {
      ...baseProps,
      elapsedTime: 1500,
      lastRunStartTimestamp: 1_704_096_000 as Seconds,
    };

    const { container } = render(
      <TooltipProvider>
        <CellStatusComponent {...props} />
      </TooltipProvider>,
    );
    expect(container).toMatchSnapshot();
  });
});
