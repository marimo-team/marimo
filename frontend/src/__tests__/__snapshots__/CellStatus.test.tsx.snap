// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`CellStatusComponent > renders completed run with elapsed time 1`] = `
<div>
  <div
    class="cell-status-icon elapsed-time hover-action"
    data-state="closed"
    data-status="idle"
    data-testid="cell-status"
  >
    <span>
      1.50s
    </span>
  </div>
</div>
`;

exports[`CellStatusComponent > renders disabled and stale state 1`] = `
<div>
  <div
    class="cell-status-icon cell-status-stale"
    data-state="closed"
    data-status="stale"
    data-testid="cell-status"
  >
    <div
      class="multi-icon relative w-fit"
    >
      <svg
        class="lucide lucide-ban h-5 w-5"
        fill="none"
        height="24"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="1.5"
        viewBox="0 0 24 24"
        width="24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <circle
          cx="12"
          cy="12"
          r="10"
        />
        <path
          d="m4.9 4.9 14.2 14.2"
        />
      </svg>
      <div
        class="second-icon absolute bottom-[-2px] right-[-2px] rounded-full"
      >
        <svg
          class="lucide lucide-refresh-cw h-3 w-3"
          fill="none"
          height="24"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2.5"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M3 12a9 9 0 0 1 9-9 9.75 9.75 0 0 1 6.74 2.74L21 8"
          />
          <path
            d="M21 3v5h-5"
          />
          <path
            d="M21 12a9 9 0 0 1-9 9 9.75 9.75 0 0 1-6.74-2.74L3 16"
          />
          <path
            d="M8 16H3v5"
          />
        </svg>
      </div>
    </div>
  </div>
</div>
`;

exports[`CellStatusComponent > renders disabled state 1`] = `
<div>
  <div
    class="cell-status-icon cell-status-disabled"
    data-state="closed"
    data-status="disabled"
    data-testid="cell-status"
  >
    <svg
      class="lucide lucide-ban h-5 w-5"
      fill="none"
      height="24"
      stroke="currentColor"
      stroke-linecap="round"
      stroke-linejoin="round"
      stroke-width="1.5"
      viewBox="0 0 24 24"
      width="24"
      xmlns="http://www.w3.org/2000/svg"
    >
      <circle
        cx="12"
        cy="12"
        r="10"
      />
      <path
        d="m4.9 4.9 14.2 14.2"
      />
    </svg>
  </div>
</div>
`;

exports[`CellStatusComponent > renders disabled transitively state 1`] = `
<div>
  <div
    class="cell-status-icon cell-status-stale"
    data-state="closed"
    data-status="disabled-transitively"
    data-testid="cell-status"
  >
    <div
      class="multi-icon relative w-fit"
    >
      <svg
        class="lucide lucide-workflow h-5 w-5"
        fill="none"
        height="24"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="1.5"
        viewBox="0 0 24 24"
        width="24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <rect
          height="8"
          rx="2"
          width="8"
          x="3"
          y="3"
        />
        <path
          d="M7 11v4a2 2 0 0 0 2 2h4"
        />
        <rect
          height="8"
          rx="2"
          width="8"
          x="13"
          y="13"
        />
      </svg>
      <div
        class="second-icon absolute top-[-2px] left-[-2px] rounded-full"
      >
        <svg
          class="lucide lucide-ban h-3 w-3"
          fill="none"
          height="24"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2.5"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <circle
            cx="12"
            cy="12"
            r="10"
          />
          <path
            d="m4.9 4.9 14.2 14.2"
          />
        </svg>
      </div>
    </div>
  </div>
</div>
`;

exports[`CellStatusComponent > renders edited state 1`] = `
<div>
  <div
    class="cell-status-icon flex items-center gap-2"
  >
    <div
      class="cell-status-stale"
      data-state="closed"
      data-status="outdated"
      data-testid="cell-status"
    >
      <svg
        class="lucide lucide-refresh-cw h-5 w-5"
        fill="none"
        height="24"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="1.5"
        viewBox="0 0 24 24"
        width="24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M3 12a9 9 0 0 1 9-9 9.75 9.75 0 0 1 6.74 2.74L21 8"
        />
        <path
          d="M21 3v5h-5"
        />
        <path
          d="M21 12a9 9 0 0 1-9 9 9.75 9.75 0 0 1-6.74-2.74L3 16"
        />
        <path
          d="M8 16H3v5"
        />
      </svg>
    </div>
    <div
      class="elapsed-time hover-action"
      data-state="closed"
      data-status="outdated"
      data-testid="cell-status"
    >
      <span>
        1.50s
      </span>
    </div>
  </div>
</div>
`;

exports[`CellStatusComponent > renders interrupted state 1`] = `
<div>
  <div
    class="cell-status-icon flex items-center gap-2"
  >
    <div
      class="cell-status-stale"
      data-state="closed"
      data-status="outdated"
      data-testid="cell-status"
    >
      <svg
        class="lucide lucide-refresh-cw h-5 w-5"
        fill="none"
        height="24"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="1.5"
        viewBox="0 0 24 24"
        width="24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M3 12a9 9 0 0 1 9-9 9.75 9.75 0 0 1 6.74 2.74L21 8"
        />
        <path
          d="M21 3v5h-5"
        />
        <path
          d="M21 12a9 9 0 0 1-9 9 9.75 9.75 0 0 1-6.74-2.74L3 16"
        />
        <path
          d="M8 16H3v5"
        />
      </svg>
    </div>
    <div
      class="elapsed-time hover-action"
      data-state="closed"
      data-status="outdated"
      data-testid="cell-status"
    >
      <span>
        1.50s
      </span>
    </div>
  </div>
</div>
`;

exports[`CellStatusComponent > renders queued state 1`] = `
<div>
  <div
    class="cell-status-icon cell-status-queued"
    data-state="closed"
    data-status="queued"
    data-testid="cell-status"
  >
    <svg
      class="lucide lucide-ellipsis h-5 w-5"
      fill="none"
      height="24"
      stroke="currentColor"
      stroke-linecap="round"
      stroke-linejoin="round"
      stroke-width="1.5"
      viewBox="0 0 24 24"
      width="24"
      xmlns="http://www.w3.org/2000/svg"
    >
      <circle
        cx="12"
        cy="12"
        r="1"
      />
      <circle
        cx="19"
        cy="12"
        r="1"
      />
      <circle
        cx="5"
        cy="12"
        r="1"
      />
    </svg>
  </div>
</div>
`;

exports[`CellStatusComponent > renders running state 1`] = `
<div>
  <div
    class="cell-status-icon elapsed-time running"
    data-state="closed"
    data-status="running"
    data-testid="cell-status"
  >
    <span>
      0ms
    </span>
  </div>
</div>
`;

exports[`CellStatusComponent > renders stale and disabled transitively state 1`] = `
<div>
  <div
    class="cell-status-icon cell-status-stale"
    data-state="closed"
    data-status="stale"
    data-testid="cell-status"
  >
    <div
      class="multi-icon relative w-fit"
    >
      <svg
        class="lucide lucide-refresh-cw h-5 w-5"
        fill="none"
        height="24"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="1"
        viewBox="0 0 24 24"
        width="24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M3 12a9 9 0 0 1 9-9 9.75 9.75 0 0 1 6.74 2.74L21 8"
        />
        <path
          d="M21 3v5h-5"
        />
        <path
          d="M21 12a9 9 0 0 1-9 9 9.75 9.75 0 0 1-6.74-2.74L3 16"
        />
        <path
          d="M8 16H3v5"
        />
      </svg>
      <div
        class="second-icon absolute bottom-[-2px] right-[-2px] rounded-full"
      >
        <svg
          class="lucide lucide-ban h-3 w-3"
          fill="none"
          height="24"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2.5"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <circle
            cx="12"
            cy="12"
            r="10"
          />
          <path
            d="m4.9 4.9 14.2 14.2"
          />
        </svg>
      </div>
    </div>
  </div>
</div>
`;

exports[`CellStatusComponent > renders stale inputs state 1`] = `
<div>
  <div
    class="cell-status-icon flex items-center gap-2"
  >
    <div
      class="cell-status-stale"
      data-state="closed"
      data-status="outdated"
      data-testid="cell-status"
    >
      <svg
        class="lucide lucide-refresh-cw h-5 w-5"
        fill="none"
        height="24"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="1.5"
        viewBox="0 0 24 24"
        width="24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M3 12a9 9 0 0 1 9-9 9.75 9.75 0 0 1 6.74 2.74L21 8"
        />
        <path
          d="M21 3v5h-5"
        />
        <path
          d="M21 12a9 9 0 0 1-9 9 9.75 9.75 0 0 1-6.74-2.74L3 16"
        />
        <path
          d="M8 16H3v5"
        />
      </svg>
    </div>
    <div
      class="elapsed-time hover-action"
      data-state="closed"
      data-status="outdated"
      data-testid="cell-status"
    >
      <span>
        1.50s
      </span>
    </div>
  </div>
</div>
`;

exports[`CellStatusComponent > renders uninstantiated state 1`] = `
<div>
  <div
    class="cell-status-icon flex items-center gap-2"
  >
    <div
      class="cell-status-stale"
      data-state="closed"
      data-status="outdated"
      data-testid="cell-status"
    >
      <svg
        class="lucide lucide-refresh-cw h-5 w-5"
        fill="none"
        height="24"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="1.5"
        viewBox="0 0 24 24"
        width="24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M3 12a9 9 0 0 1 9-9 9.75 9.75 0 0 1 6.74 2.74L21 8"
        />
        <path
          d="M21 3v5h-5"
        />
        <path
          d="M21 12a9 9 0 0 1-9 9 9.75 9.75 0 0 1-6.74-2.74L3 16"
        />
        <path
          d="M8 16H3v5"
        />
      </svg>
    </div>
  </div>
</div>
`;

exports[`ElapsedTime component > renders elapsed time correctly 1`] = `
<div>
  <span
    class="tracking-wide font-semibold"
  >
    1.50s
  </span>
</div>
`;
